draft-ietf-quic-transport.md

```diff
@@ -1920,10 +1920,12 @@ that are uniquely attributed to a single connection. This includes datagrams
 that contain packets that are successfully processed and datagrams that contain
 packets that are all discarded.
 
-Clients MUST ensure that UDP datagrams containing Initial packets have UDP
-payloads of at least 1200 bytes, adding PADDING frames as necessary.
-A client that sends padded datagrams allows the server to
-send more data prior to completing address validation.
+Clients MUST and servers SHOULD ensure that UDP datagrams containing Initial
+packets have UDP payloads of at least 1200 bytes, adding PADDING frames as
+necessary.  A client that sends padded datagrams allows the server to send more
```

Comment:
Maybe it would be better to limit the scope and say _servers SHOULD pad Initial packets that carry crypto data_.

The problem of padding all Initial packets is that ACK-only Initial packets consume full MTU, when the send window is still controlled by the amplification limit.

Reply:
I think making a carve-out for ACK-only packets sounds attractive, but it could result in the client sending an Initial, getting it acknowledged, but not receiving the server's Initial.  Then things get complex to reason about.

Reply:
Endpoints that don't coalesce are already trading performance for simplicity. The added sacrifice here isn't that large.

1) Clients: if it only receives an Initial, padding is critical to raising the amplification limit. If it gets a Handshake packet too, it will momentarily send a Handshake ACK that will allow it to free the Initial context and get the cwnd back.

2) Servers: If the client has a valid Retry or NEW_TOKEN token, there is no amplification limit. So the performance penalty only applies to a non-coalescing server when:
(a) There is no token, AND the Initial padding causes the handshake flight to go over the amplification limit.
(b) The initial padding causes the handshake flight to go over the initial cwnd.
I find it hard to believe a server deeply concerned about these losses would not coalesce packets for other reasons.